src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -415,10 +418,22 @@ internal SynthesizedAttributeData SynthesizeDebuggerStepThroughAttribute()
             return TrySynthesizeAttribute(WellKnownMember.System_Diagnostics_DebuggerStepThroughAttribute__ctor);
         }
 
-        internal SynthesizedAttributeData SynthesizeReadOnlyAttribute()
+        internal void EnsureIsReadOnlyAttributeExists(Symbol readOnlySymbol)
         {
-            // PROTOTYPE(readonlyRefs) it is optional now as it will be generated in the next PR
-            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_ReadOnlyAttribute__ctor, isOptionalUse: true);
+            // This will report a DeclarationDiagnostics error if the member is not found.
+            // We only need to report this when compiling NetModules, as for other compilations, we will generate it.
+
+            var attributeSymbol = Binder.GetWellKnownTypeMember(
+                compilation: this,
+                member: WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor,
+                diagnostics: DeclarationDiagnostics,
```

Comment:
Moving to a model where diagnosticsbag is defined by the syntax needing it whenever possible.

---
In reply to: [113538171](https://github.com/dotnet/roslyn/pull/18715#discussion_r113538171) [](ancestors = 113538171)